### PR TITLE
E2e cv blog post part 2

### PR DESCRIPTION
**Pull request by alex000kim**
Part 2, depends on https://github.com/iterative/dvc.org/pull/3482

Target publish date: May 4rd

- [x] Still waiting on Header image corrections from the graphic designer.
- [ ] Need to link to/from Part 1 and Part 2
- [x]  There are no grammatical errors, jargon, or obvious stylistic issues. (For reviewer)
- [x]  Have images been run through tiny.png?
- [x]  Are Oxford commas observed?
- [x]  Content Creator has done a complete review for typos/language after taking a break from content
- [x]  links checked in the piece
- [x]  Have reviewed what the piece looks like on your own machine using `yarn develop`

### CODE REVIEW

**Review by casperdcl — COMMENTED** [review]
odd this seems to contain #3482

**Comment by alex000kim** [comment]
Yeah, I branched off of the pt1 branch. That's on purpose so that one could also see pt1 when reviewing pt2.

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  exp                       Apr 09, 2022      0.13305     0.087599          0.77803         0.66494      0.89084     0.70534    0.97891   0.01                  8                  6c513ae
+  ├── 83a4975 [exp-2d80e]   Apr 09, 2022      0.11189     0.088695          0.86905         0.75296      0.92005     0.70612    0.97689   0.01                  16                 5854528
+  ├── 675efb3 [exp-6c274]   Apr 09, 2022      0.10356     0.069076          0.90321         0.75906      0.92371     0.71492    0.98099   0.1                   16                 770745a
+  └── c8b1857 [exp-04bcd]   Apr 09, 2022      0.11189     0.088695          0.86905         0.75296      0.92005     0.71619    0.98025   0.01                  8                  094c420
+ ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
+```
+
+Then we’d pick the best one (e.g. highest `dice_mean` score) and commit it to
+our git repo:
```
Comment (daavoo) [inline content/blog/2022-05-04-e2e-cv-pt2.md:111]: I think wording here might be misleading (commit has a meaning in git context that doesn't match what exp push do).

Not sure how to phrase it, but we could explicitly use the words persist and share, with links to respective docs pages (https://dvc.org/doc/user-guide/experiment-management/persisting-experiments).

Comment (daavoo) [reply content/blog/2022-05-04-e2e-cv-pt2.md:111]: And perhaps only to clarify (as people might not be aware) also explicitly mention that there is no requirement for picking a single experiment, multiple experiments can be shared / persisted

Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: Good point! Updated.

**Comment by jorgeorpinel** [comment]
>  Need to link to Part 1

to and from 🙂

**Comment by jorgeorpinel** [comment]
> >  Need to link to Part 1
>
> Yeah, I branched off of the pt1 branch.

You could make `e2e-cv-blog-part1` the base of this PR for now (mark it as draft) so it's easier to review (cascading PRs).

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+adding failed experiments to your git history. Check out this
+[blog post](https://dvc.org/blog/ml-experiment-versioning), which talks about
+the idea of "ML Experiments as Code." That means treating experiments as you'd
+treat code, that is, use git to track all changes in configs, metrics, and data
+versions through text files. This approach removes the need for a separate
+database/online service to store experiment metadata. If we run a few
+experiments:
+
+```bash
+$ dvc exp run --set-param train.learning_rate=0.1
```
Comment (daavoo) [inline content/blog/2022-05-04-e2e-cv-pt2.md:87]: Maybe add a Link or Mention for the queue option https://dvc.org/doc/user-guide/experiment-management/running-experiments#the-experiments-queue ?

Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md:87]: yup, done

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+```
+
+or we can share them with our teammates by
+[uploading](https://dvc.org/doc/user-guide/experiment-management/sharing-experiments#uploading-experiments)
+them to our git repo:
+
+```bash
+$ dvc exp push origin exp-04bcd
+Pushed experiment 'exp-04bcd'to Git remote 'origin'.
+```
```
Comment (dberenbaum) [inline content/blog/2022-05-04-e2e-cv-pt2.md:136]: Not sure if it's worth including `exp push` here given the mention of Studio below. They don't really integrate well at the moment, so it might be more confusing than useful.

Comment (dberenbaum) [reply content/blog/2022-05-04-e2e-cv-pt2.md:136]: Related: https://github.com/iterative/studio-support/issues/45

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+There is a super convenient set of
+[Experiment Management](https://dvc.org/doc/user-guide/experiment-management)
+features that make switching between reproducible experiments very easy without
+adding failed experiments to your git history. Check out this
+[blog post](https://dvc.org/blog/ml-experiment-versioning), which talks about
+the idea of "ML Experiments as Code." That means treating experiments as you'd
+treat code, that is, use git to track all changes in configs, metrics, and data
+versions through text files. This approach removes the need for a separate
+database/online service to store experiment metadata. If we run a few
+experiments:
```
Comment (dberenbaum) [inline content/blog/2022-05-04-e2e-cv-pt2.md:84]: It seems like there's something missing after "If we run a few experiments"? Then what?

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+```bash
+$ dvc exp run --set-param train.learning_rate=0.1
+...
+$ dvc exp run --set-param train.learning_rate=0.01
+...
+$ dvc exp run --set-param train.batch_size=16
+...
+```
+
+Optionally, you can create delay the execution of the experiments by putting
```
Comment (dberenbaum) [inline content/blog/2022-05-04-e2e-cv-pt2.md:95]:
```suggestion
Optionally, you can delay the execution of the experiments by putting
```

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+[templating format](https://dvc.org/doc/user-guide/project-structure/pipelines-files#templating)
+to insert values from different sources in the YAML structure itself.
+
+DVC tracks which stages of the pipeline experienced changes and only reruns
+those. By changes, we mean _everything_ that might affect the predictive
+performance of your model like changes to the dataset, source code and/or
+parameters. This not only ensures complete reproducibility but often
+significantly reduces the time needed to rerun the whole pipeline while ensuring
+consistent results on every rerun. For example, if we only change the code in
+the evaluation stage (e.g., change what type of metrics we want to compute),
```
Comment (dberenbaum) [inline content/blog/2022-05-04-e2e-cv-pt2.md:59]: It might be nice to use a real example here, like why you added `dice_mean` and how it differs from other `acc_mean` or other metrics you are capturing.

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+features that make switching between reproducible experiments very easy without
+adding failed experiments to your git history. Check out this
+[blog post](https://dvc.org/blog/ml-experiment-versioning), which talks about
+the idea of "ML Experiments as Code." That means treating experiments as you'd
+treat code, that is, use git to track all changes in configs, metrics, and data
+versions through text files. This approach removes the need for a separate
+database/online service to store experiment metadata. If we run a few
+experiments:
+
+```bash
```
Comment (dberenbaum) [inline content/blog/2022-05-04-e2e-cv-pt2.md:86]: Here's another chance to tie it back to the use case. Did you see something from the previous experiment that made you decide to adjust the learning rate or batch size?

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+In the screenshot below, you can see that we created a separate `exp` branch
+that displays the results of the local experiments that we decided to upload to
+our remote repository, like trying different learning rates and batch sizes.
+Note that earlier, we discarded all local experiments whose performance we
+weren't satisfied with.
+
+![Studio view](/uploads/images/2022-05-04/studio_view.png '=800')
+
+To track detailed information about the training process, we integrated
```
Comment (dberenbaum) [inline content/blog/2022-05-04-e2e-cv-pt2.md:166]: Should this be a new section? It doesn't seem related to collaboration or Studio.

Comment (dberenbaum) [reply content/blog/2022-05-04-e2e-cv-pt2.md:166]: And should it be earlier in the blog post? The example shows results in Studio, but the use case you mention is about diagnosing issues during local training, which would happen before pushing to Studio, right?

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 3 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: removed

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 4 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: yup, fixed

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 6 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: Clarified and linked to a blog post on image segmentation metrics.

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 7 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md:92]: rephrased

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 8 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: moved to the earlier section

**Comment by jendefig** [comment]
How come @alex000kim's posts aren't deploying on the Gatsby thing?  cc: @rogermparent

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
+---
+title: 'End-to-End Computer Vision API, Part 2: Local Experiments'
+date: 2022-05-04
+description: >
+  In part 1 of this series of posts, we introduced a solution to a common
+  problem faced by companies in the manufacturing industry: detecting defects
+  from images of products moving along a production line. The solution we
+  proposed was a Deep Learning-based image segmentation model wrapped in a web
+  API. We talked about effective management and versioning of large datasets and
+  the creation of reproducible ML pipelines.
```
Comment (jendefig) [inline content/blog/2022-05-04-e2e-cv-pt2.md:10]: But what will we do in this part?  You could summarize what was in first part and add what will happen in Part 2.

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+descriptionLong: |
+  _This is the second part of a three-part series of posts._
+
+  In part 1 of this series of posts, we introduced a solution to a common problem faced by companies in the manufacturing industry: detecting defects from images of products moving along a production line. The solution we proposed was a Deep Learning-based image segmentation model wrapped in a web API. We talked about effective management and versioning of large datasets and the creation of reproducible ML pipelines.
```
Comment (jendefig) [inline content/blog/2022-05-04-e2e-cv-pt2.md:15]: Same issue as short description.

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+picture: 2022-05-04/e2e-cv-pt2-cover.png
+author: alex_kim
+commentsUrl: https://discuss.dvc.org/t/end-to-end-computer-vision/1178
```
Comment (jendefig) [inline content/blog/2022-05-04-e2e-cv-pt2.md:19]: Wondering if we should have a separate topic for each?  Not sure it's necessary though since this is an interconnected series.  I think it's ok to leave like this.  We rarely get comments on blog posts.

File: content/blog/2022-05-04-e2e-cv-pt2.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+function throughout training (enabled by the earlier integration of
+[DVCLive](https://dvc.org/doc/dvclive)) for a set of selected commits.
+
+![DVCLive metrics displayed in
+Studio](/uploads/images/2022-05-04/dvc_live_studio.png '=800')
+
+Now, for example, if we see that the loss function hasn't reached a plateau
+after a certain number of epochs, we'll try increasing this number. Or, even
+worse, if we see the loss function growing over time, it'll be an indication
+that our learning rate may be too high.
```
Comment (jendefig) [inline content/blog/2022-05-04-e2e-cv-pt2.md:179]: I feel there needs to be a concluding sentence in this paragraph that leads back to the point of this post like, "So it will be back to the experimentation drawing board." or something like that.  Feels abrupt before jumping to the summary as is.

**Review by jendefig — CHANGES_REQUESTED** [review]
Also looks great!  Just a few minor things.

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 9 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: Paraphrased this because this short description gets truncated and you can't see it in full anyway.

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 10 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: Added a summary for this pt2 post

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 11 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md:26]: Since all parts are closely related, I think it's better to have one place for potential comments.

File: content/blog/2022-05-04-e2e-cv-pt2.md (hunk 12 above)
Comment (alex000kim) [reply content/blog/2022-05-04-e2e-cv-pt2.md]: Yup, added a concluding sentence.

**Comment by alex000kim** [comment]
> How come @alex000kim's posts aren't deploying on the Gatsby thing? cc: @rogermparent

I am guessing it's because the target branch right now is not `master`

**Comment by rogermparent** [comment]
> I am guessing it's because the target branch right now is not `master`

That's correct, Gatsby Cloud only creates deploy previews for PRs targeting the main branch the site is configured with. I got around it for MLEM docs by making another website entry with the main branch being the one that adds a docs engine.

**Comment by alex000kim** [comment]
last deployment: https://build-04b6ee7c-a4bc-4989-aaf1-a2ac6f6024c2.gtsb.io/blog/e2e-cv-pt2

File: content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+  best-performing ones.
+- Track the global history of the model's performance, and map each improvement
+  to a particular change in code, configuration, or data.
+- Zoom into the details of each training run to help us diagnose issues.
+
+### Experiment Management
+
+Our DVC pipeline relies on the parameters defined in the
+[`params.yaml`](https://github.com/iterative/magnetic-tiles-defect/blob/main/params.yaml)
+file. By loading its contents in each stage, we can avoid hard-coded parameters.
```
Comment (daavoo) [inline content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md:53]: Perhaps it's my understanding but it reads to me like DVC *requires* this file (because of *relies*) and somehow implies that DVC only supports `params.yaml` for configuration, when it supports arbitrary numbers and names of parameters files (from https://dvc.org/doc/command-reference/params#description):

> The default parameters file name is params.yaml, but any other YAML 1.2, JSON, TOML, or [Python](https://dvc.org/doc/command-reference/params#examples-python-parameters-file) files can be used additionally (listed under params: with a sub-list of param values, as shown in the sample above) .

Sorry for late comment 🙏 Not critical, can be released without addressing

**Review by daavoo — APPROVED** [review]

File: content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md (hunk 13 above)
Comment (jendefig) [reply content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md:53]: Good point @daavoo!  Maybe should change to 

"Our DVC pipeline relies on the parameters defined in the[`params.yaml]( https://github.com/iterative/magnetic-tiles-defect/blob/main/params.yaml) file in this case (see [other possible file types here]( https://dvc.org/doc/command-reference/params#description)."

That way it keeps readers on track, but if they are interested in other types they know it and can get info on it.

Comment (alex000kim) [reply content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md]: Good point! Updated with @jendefig's suggestion.

File: content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+
+Optionally, you can delay the execution of the experiments by putting them in a
+[queue](https://dvc.org/doc/user-guide/experiment-management/running-experiments#the-experiments-queue),
+and execute them later with the `dvc exp run --run-all` command.
+
+These local experiments are powered by Git references, and you can learn about
+them in [this post](https://dvc.org/blog/experiment-refs). We can display all
+experiments with the `dvc exp show` command:
+
+```bash
```
Comment (shcheklein) [inline content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md:119]: we have `dvctable` that can highlight the table nicely

File: content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+  └── c8b1857 [exp-04bcd]   Apr 09, 2022      0.11189     0.088695          0.86905         0.75296      0.92005     0.71619    0.98025   0.01                  8                  094c420
+ ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
+```
+
+Once we identify one or a few best ones (e.g., highest `dice_mean` score), we
+can
+[persist](https://dvc.org/doc/user-guide/experiment-management/persisting-experiments)
+them by creating a branch out of an experiment:
+
+```bash
```
Comment (shcheklein) [inline content/blog/2022-05-05-end-to-end-computer-vision-api-part-2-local-experiments.md:137]: it should be `dvc`, not `bash`
